### PR TITLE
status: wrap status proto in a struct

### PR DESCRIPTION
**Pull request by zouyee**
fix https://github.com/grpc/grpc-go/issues/3536

Signed-off-by: Zou Nengren <zouyee1989@gmail.com>

### CODE REVIEW

File: internal/status/status.go [hunk 1]
```diff
@@ -143,16 +145,20 @@ func (s *Status) Details() []interface{} {
 
 // Error is an alias of a status proto. It implements error and Status,
 // and a nil Error should never be returned by this package.
-type Error spb.Status
+type Error struct {
+	e spb.Status
```
Comment (easwars) [inline internal/status/status.go:149]: I think we could simply store a pointer to `spb.Status` here for a couple of reasons. When `Error` is created in `Status.Err`, the proto wrapped in `Status` is cloned before passing it on to `Error`. Similarly, in `Error.GRPCStatus`, the proto is cloned. So, we can avoid another copy by storing a pointer here.

File: internal/status/status.go [hunk 2]
```diff
@@ -143,16 +145,20 @@ func (s *Status) Details() []interface{} {
...
+}
+
+var _ = error(&Error{})
 
 func (se *Error) Error() string {
-	p := (*spb.Status)(se)
+	p := se.e
```
Comment (easwars) [inline internal/status/status.go:155]: We can probably get rid of the local variable `p` here.

File: internal/status/status.go [hunk 3]
```diff
@@ -143,16 +145,20 @@ func (s *Status) Details() []interface{} {
 
 // Error is an alias of a status proto. It implements error and Status,
 // and a nil Error should never be returned by this package.
-type Error spb.Status
+type Error struct {
+	e spb.Status
+}
+
+var _ = error(&Error{})
```
Comment (easwars) [inline internal/status/status.go:152]: Ditto about not needing this.

**Comment by easwars** [comment]
@dfawley : Would it make sense to run benchmarks for this change?

File: internal/status/status.go [hunk 4]
```diff
@@ -143,16 +145,20 @@ func (s *Status) Details() []interface{} {
 
 // Error is an alias of a status proto. It implements error and Status,
```
Comment (dfawley) [inline internal/status/status.go:146]: This comment would no longer be accurate.

File: internal/status/status.go [hunk 5]
```diff
@@ -43,6 +43,8 @@ type Status struct {
 	s *spb.Status
 }
 
+var _ = error(&Status{})
```
[the remark this answers is not in the document]
Comment (dfawley) [reply internal/status/status.go:46]: :open_mouth: this is bad.  Status _cannot_ implement error.  "OK" Statuses are valid, and non-nil.  I'll send a change to remove the Error method on Status ASAP.

[the remark this answers is not in the document]
Comment (dfawley) [reply internal/status/status.go:46]: Sent #3561.

**Comment by dfawley** [comment]
> @dfawley : Would it make sense to run benchmarks for this change?

I'm not too concerned about performance here.  For a few reasons:

1. These are errors, and errors aren't generally optimized.
2. None of our benchmarks have error-producing paths, so it would need a microbenchmark to be written to measure.
3. It's unlikely this kind of change would cause a performance regression relevant for RPCs.

**Comment by easwars** [comment]
https://github.com/grpc/grpc-go/pull/3561 has been merged.

Please do take a look. Thanks.

**Comment by zouyee** [comment]
👌

**Comment by zouyee** [comment]
OK

**Comment by menghanl** [comment]
Friendly ping

**Comment by zouyee** [comment]
@easwars  @menghanl PTAL

**Comment by dfawley** [comment]
@easwars were all your comments addressed?  If so LGTM.

**Pull request merged**